internal/reconcile_components.go

```diff
@@ -0,0 +1,126 @@
+package internal
+
+import (
+	"fmt"
+	"github.com/broadinstitute/terra-status-manager/internal/shared"
+	"github.com/broadinstitute/terra-status-manager/internal/statuspage"
+	"github.com/broadinstitute/terra-status-manager/pkg"
+	"github.com/go-resty/resty/v2"
+	"github.com/mitchellh/mapstructure"
+	"reflect"
+)
+
+// componentsToDelete provides a slice of remote components that don't correlate to an entry in the config
+func componentsToDelete(
+	configComponentMap map[string]pkg.Component,
+	remoteComponentMap map[string]statuspage.Component,
+) []statuspage.Component {
+	var ret []statuspage.Component
+	for name, remoteComponent := range remoteComponentMap {
+		if _, found := configComponentMap[name]; !found {
+			ret = append(ret, remoteComponent)
+		}
+	}
+	return ret
+}
+
+// componentsToCreate provides a slice of components that aren't present on the remote
+func componentsToCreate(
+	configComponentMap map[string]pkg.Component,
+	remoteComponentMap map[string]statuspage.Component,
+) ([]statuspage.Component, error) {
+	var ret []statuspage.Component
+	for name, configComponent := range configComponentMap {
+		if _, found := remoteComponentMap[name]; !found {
+			var c statuspage.Component
+			err := statuspage.ComponentConfigToApi(configComponent, &c)
+			if err != nil {
+				return nil, fmt.Errorf("error decoding pkg.Component to statuspage.Component: %w", err)
+			}
+			// We specifically don't want status to be influenced by config file; components start out operational
+			c.Status = "operational"
+			ret = append(ret, c)
+		}
+	}
+	return ret, nil
+}
+
+// componentsToModify provides a slice of components that should be modified on the remote
+func componentsToModify(
+	configComponentMap map[string]pkg.Component,
+	remoteComponentMap map[string]statuspage.Component,
+) ([]statuspage.Component, error) {
+	var ret []statuspage.Component
+	for name, configComponent := range configComponentMap {
+		if remoteComponent, found := remoteComponentMap[name]; found {
+			var c statuspage.Component
+			err := mapstructure.Decode(remoteComponent, &c)
```

Comment:
I'm a bit hesitant to do that because it doesn't reduce code length at all--no matter what we will still have
```go
err = call(…)
if err != nil {
    return nil, err
}
```
I think the indirection might be more difficult to maintain over the long run--GoLand is happy to help with syntax highlighting and checking format strings but as soon as we start passing them around it'll get confused. I'll be the first to admit that there is nothing clever about how I'm handling errors but I thought maybe it being dead simple would be worth it.